### PR TITLE
UnnecessaryAbstractClass: Support exclusion via Annotations.

### LINKED ISSUE #377: UnnecessaryAbstractClass: Relax when the class is annotated with Module
In dagger it makes sense to have an abstract class with a few abstract methods since it'll tell the Dagger Compiler how certain types can be bound.

An example is the following class:

```kotlin
@Module abstract class SomeModule {
  @Binds abstract fun provideSomeThing(mySomeThing: MySomeThing): SomeThing
}
```

We probably want to ignore those cases when `dagger.Module` is applied.

What do you think?

### PR DESCRIPTION
**Pull request by vanniktech**
Fixes #377

### CODE REVIEW

File: detekt-cli/src/main/resources/default-detekt-config.yml [hunk 1]
```diff
@@ -473,6 +473,7 @@ style:
     active: false
   UnnecessaryAbstractClass:
     active: false
+    excludeAnnotatedClasses: "dagger.Module"
```
Comment (schalkms) [inline detekt-cli/src/main/resources/default-detekt-config.yml]: I think `excludeAnnotatedClasses` should per default be empty.
Not everyone uses this framework.

File: detekt-rules/build.gradle.kts [hunk 2]
```diff
@@ -14,6 +14,7 @@ dependencies {
 	implementation(project(":detekt-api"))
 
 	testImplementation(project(":detekt-test"))
+  testImplementation("com.google.dagger:dagger:2.17")
```
Comment (schalkms) [inline detekt-rules/build.gradle.kts]: Why should detekt have a dependency to this framework? Is there any usage?

File: detekt-rules/src/test/resources/cases/UnnecessaryAbstractClassNegative.kt [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 package cases
 
+import dagger.Module
```
Comment (schalkms) [inline detekt-rules/src/test/resources/cases/UnnecessaryAbstractClassNegative.kt]: Ah here's the usage of dagger.
Please use an annotation in the stdlib for the test case or a custom one.

File: detekt-cli/src/main/resources/default-detekt-config.yml (hunk 1 above)
Comment (vanniktech) [reply detekt-cli/src/main/resources/default-detekt-config.yml]: In my opinion, Detekt should have good default values. In 99% of the cases Dagger modules get flagged from Detekt. If we have it as a default it won't be flagged and less configuration is needed.

File: detekt-rules/build.gradle.kts (hunk 2 above)
Comment (vanniktech) [reply detekt-rules/build.gradle.kts]: Yup in the Case file.

File: detekt-rules/src/test/resources/cases/UnnecessaryAbstractClassNegative.kt (hunk 3 above)
Comment (vanniktech) [reply detekt-rules/src/test/resources/cases/UnnecessaryAbstractClassNegative.kt]: Check https://github.com/arturbosch/detekt/pull/1091#discussion_r215885103 first please. I can still revisit.

**Comment by vanniktech** [comment]
> Please remove the dagger dependency to make this mergeable.

It's only applicable for tests so it should not be problem.

**Comment by schalkms** [comment]
Anyways, adding a (potentially huge) framework just for 1 annotation test case doesn't make much sense.

**Comment by vanniktech** [comment]
So if I change the annotation that we will use for testing this is good to go?

**Comment by schalkms** [comment]
Yep! The `excludeAnnotatedClasses` property is discussable IMHO.

**Comment by vanniktech** [comment]
Adjusted to your feedback. @Mauin what do you think about having a default for Dagger?

**Pull request merged**